File: benchmarks/runtime/gate-b/nbody-languages/sources/nbody.lua
-- The Computer Language Benchmarks Game
-- https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/nbody-lua-2.html
-- contributed by Mike Pall; modified by Geoff Leyland
-- Adapted only to time the steady-state advance loop and emit CSV.

local sqrt = math.sqrt
local PI = 3.141592653589793
local SOLAR_MASS = 4 * PI * PI
local DAYS_PER_YEAR = 365.24
local bodies = {
  {
    x = 0, y = 0, z = 0,
    vx = 0, vy = 0, vz = 0,
    mass = SOLAR_MASS
  },
  {
    x = 4.84143144246472090e+00,
    y = -1.16032004402742839e+00,
    z = -1.03622044471123109e-01,
    vx = 1.66007664274403694e-03 * DAYS_PER_YEAR,
    vy = 7.69901118419740425e-03 * DAYS_PER_YEAR,
    vz = -6.90460016972063023e-05 * DAYS_PER_YEAR,
    mass = 9.54791938424326609e-04 * SOLAR_MASS
  },
  {
    x = 8.34336671824457987e+00,
    y = 4.12479856412430479e+00,
    z = -4.03523417114321381e-01,
    vx = -2.76742510726862411e-03 * DAYS_PER_YEAR,
    vy = 4.99852801234917238e-03 * DAYS_PER_YEAR,
    vz = 2.30417297573763929e-05 * DAYS_PER_YEAR,
    mass = 2.85885980666130812e-04 * SOLAR_MASS
  },
  {
    x = 1.28943695621391310e+01,
    y = -1.51111514016986312e+01,
    z = -2.23307578892655734e-01,
    vx = 2.96460137564761618e-03 * DAYS_PER_YEAR,
    vy = 2.37847173959480950e-03 * DAYS_PER_YEAR,
    vz = -2.96589568540237556e-05 * DAYS_PER_YEAR,
    mass = 4.36624404335156298e-05 * SOLAR_MASS
  },
  {
    x = 1.53796971148509165e+01,
    y = -2.59193146099879641e+01,
    z = 1.79258772950371181e-01,
    vx = 2.68067772490389322e-03 * DAYS_PER_YEAR,
    vy = 1.62824170038242295e-03 * DAYS_PER_YEAR,
    vz = -9.51592254519715870e-05 * DAYS_PER_YEAR,
    mass = 5.15138902046611451e-05 * SOLAR_MASS
  }
}

local function advance(system, nbody, dt)
  for i = 1, nbody do
    local bi = system[i]
    local bix, biy, biz, bimass = bi.x, bi.y, bi.z, bi.mass
    local bivx, bivy, bivz = bi.vx, bi.vy, bi.vz
    for j = i + 1, nbody do
      local bj = system[j]
      local dx, dy, dz = bix - bj.x, biy - bj.y, biz - bj.z
      local mag = sqrt(dx * dx + dy * dy + dz * dz)
      mag = dt / (mag * mag * mag)
      local bm = bj.mass * mag
      bivx = bivx - dx * bm
      bivy = bivy - dy * bm
      bivz = bivz - dz * bm
      bm = bimass * mag
      bj.vx = bj.vx + dx * bm
      bj.vy = bj.vy + dy * bm
      bj.vz = bj.vz + dz * bm
    end
    bi.vx, bi.vy, bi.vz = bivx, bivy, bivz
    bi.x = bix + dt * bivx
    bi.y = biy + dt * bivy
    bi.z = biz + dt * bivz
  end
end

local function energy(system, nbody)
  local e = 0
  for i = 1, nbody do
    local bi = system[i]
    local vx, vy, vz, bim = bi.vx, bi.vy, bi.vz, bi.mass
    e = e + 0.5 * bim * (vx * vx + vy * vy + vz * vz)
    for j = i + 1, nbody do
      local bj = system[j]
      local dx, dy, dz = bi.x - bj.x, bi.y - bj.y, bi.z - bj.z
      local distance = sqrt(dx * dx + dy * dy + dz * dz)
      e = e - bim * bj.mass / distance
    end
  end
  return e
end

local function offset_momentum(system, nbody)
  local px, py, pz = 0, 0, 0
  for i = 1, nbody do
    local bi = system[i]
    local bim = bi.mass
    px = px + bi.vx * bim
    py = py + bi.vy * bim
    pz = pz + bi.vz * bim
  end
  system[1].vx = -px / SOLAR_MASS
  system[1].vy = -py / SOLAR_MASS
  system[1].vz = -pz / SOLAR_MASS
end

local turns = tonumber(arg and arg[1]) or 1000000
local implementation = (arg and arg[2]) or "lua-game-2"
local nbody = #bodies
offset_momentum(bodies, nbody)
local initial_energy = energy(bodies, nbody)
local warmup = {}
for i = 1, nbody do
  local body = bodies[i]
  warmup[i] = {
    x = body.x, y = body.y, z = body.z,
    vx = body.vx, vy = body.vy, vz = body.vz,
    mass = body.mass
  }
end
for _ = 1, 1000 do
  advance(warmup, nbody, 0.01)
end
collectgarbage("collect")
local memory_before = collectgarbage("count") * 1024
local started = os.clock()
for _ = 1, turns do
  advance(bodies, nbody, 0.01)
end
local seconds = os.clock() - started
local memory_after = collectgarbage("count") * 1024
local final_energy = energy(bodies, nbody)
io.write(string.format(
  "%s,%d,%.9f,%.3f,%.3f,%.12f,%.12f,0,%.0f\n",
  implementation, turns, seconds, seconds * 1e9 / turns,
  turns / seconds, initial_energy, final_energy, memory_after - memory_before
))

File: benchmarks/runtime/gate-b/nbody-languages/sources/nbody.rs
// The Computer Language Benchmarks Game
// https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/nbody-rust-3.html
// Contributed by Ilia Schelokov.
// Adapted only to time the steady-state advance loop and emit CSV.

use std::default::Default;
use std::f64::consts::PI;
use std::ops::{Add, AddAssign, Mul, Sub, SubAssign};
use std::time::Instant;

#[derive(Clone, Debug)]
struct Vec3D(f64, f64, f64);

impl Vec3D {
    fn sum_squares(&self) -> f64 {
        self.0 * self.0 + self.1 * self.1 + self.2 * self.2
    }

    fn magnitude(&self, dt: f64) -> f64 {
        let sum = self.sum_squares();
        dt / (sum * sum.sqrt())
    }
}

impl Default for Vec3D {
    fn default() -> Vec3D {
        Vec3D(0.0, 0.0, 0.0)
    }
}

impl Add for &Vec3D {
    type Output = Vec3D;

    fn add(self, rhs: Self) -> Self::Output {
        Vec3D(self.0 + rhs.0, self.1 + rhs.1, self.2 + rhs.2)
    }
}

impl Sub for &Vec3D {
    type Output = Vec3D;

    fn sub(self, rhs: Self) -> Self::Output {
        Vec3D(self.0 - rhs.0, self.1 - rhs.1, self.2 - rhs.2)
    }
}

impl Mul<f64> for &Vec3D {
    type Output = Vec3D;

    fn mul(self, rhs: f64) -> Self::Output {
        Vec3D(self.0 * rhs, self.1 * rhs, self.2 * rhs)
    }
}

impl AddAssign for Vec3D {
    fn add_assign(&mut self, rhs: Self) {
        self.0 += rhs.0;
        self.1 += rhs.1;
        self.2 += rhs.2;
    }
}

impl SubAssign for Vec3D {
    fn sub_assign(&mut self, rhs: Self) {
        self.0 -= rhs.0;
        self.1 -= rhs.1;
        self.2 -= rhs.2;
    }
}

#[derive(Clone, Debug)]
struct Body {
    position: Vec3D,
    velocity: Vec3D,
    mass: f64,
}

const BODIES_COUNT: usize = 5;
const SOLAR_MASS: f64 = 4.0 * PI * PI;
const DAYS_PER_YEAR: f64 = 365.24;
const INTERACTIONS: usize = BODIES_COUNT * (BODIES_COUNT - 1) / 2;

const STARTING_STATE: [Body; BODIES_COUNT] = [
    Body {
        mass: SOLAR_MASS,
        position: Vec3D(0.0, 0.0, 0.0),
        velocity: Vec3D(0.0, 0.0, 0.0),
    },
    Body {
        position: Vec3D(
            4.841_431_442_464_72,
            -1.160_320_044_027_428_4,
            -1.036_220_444_711_231_1e-1,
        ),
        velocity: Vec3D(
            1.660_076_642_744_037e-3 * DAYS_PER_YEAR,
            7.699_011_184_197_404e-3 * DAYS_PER_YEAR,
            -6.904_600_169_720_63e-5 * DAYS_PER_YEAR,
        ),
        mass: 9.547_919_384_243_266e-4 * SOLAR_MASS,
    },
    Body {
        position: Vec3D(
            8.343_366_718_244_58,
            4.124_798_564_124_305,
            -4.035_234_171_143_214e-1,
        ),
        velocity: Vec3D(
            -2.767_425_107_268_624e-3 * DAYS_PER_YEAR,
            4.998_528_012_349_172e-3 * DAYS_PER_YEAR,
            2.304_172_975_737_639_3e-5 * DAYS_PER_YEAR,
        ),
        mass: 2.858_859_806_661_308e-4 * SOLAR_MASS,
    },
    Body {
        position: Vec3D(
            1.289_436_956_213_913_1e1,
            -1.511_115_140_169_863_1e1,
            -2.233_075_788_926_557_3e-1,
        ),
        velocity: Vec3D(
            2.964_601_375_647_616e-3 * DAYS_PER_YEAR,
            2.378_471_739_594_809_5e-3 * DAYS_PER_YEAR,
            -2.965_895_685_402_375_6e-5 * DAYS_PER_YEAR,
        ),
        mass: 4.366_244_043_351_563e-5 * SOLAR_MASS,
    },
    Body {
        position: Vec3D(
            1.537_969_711_485_091_1e1,
            -2.591_931_460_998_796_4e1,
            1.792_587_729_503_711_8e-1,
        ),
        velocity: Vec3D(
            2.680_677_724_903_893_2e-3 * DAYS_PER_YEAR,
            1.628_241_700_382_423e-3 * DAYS_PER_YEAR,
            -9.515_922_545_197_159e-5 * DAYS_PER_YEAR,
        ),
        mass: 5.151_389_020_466_114_5e-5 * SOLAR_MASS,
    },
];

fn advance(bodies: &mut [Body; BODIES_COUNT], dt: f64, steps: usize) {
    let mut d_positions: [Vec3D; INTERACTIONS] = Default::default();
    let mut magnitudes = [0.0; INTERACTIONS];
    for _ in 0..steps {
        let mut k = 0;
        for (i, body1) in bodies.iter().enumerate() {
            for body2 in &bodies[i + 1..] {
                d_positions[k] = &body1.position - &body2.position;
                k += 1;
            }
        }
        for (mag, d_pos) in magnitudes.iter_mut().zip(d_positions.iter()) {
            *mag = d_pos.magnitude(dt);
        }
        let mut k = 0;
        for i in 0..BODIES_COUNT - 1 {
            let (body1, rest) = bodies[i..].split_first_mut().unwrap();
            for body2 in rest {
                let d_pos = &d_positions[k];
                let mag = magnitudes[k];
                body1.velocity -= d_pos * (body2.mass * mag);
                body2.velocity += d_pos * (body1.mass * mag);
                k += 1;
            }
        }
        for body in bodies.iter_mut() {
            body.position += &body.velocity * dt;
        }
    }
}

fn offset_momentum(bodies: &mut [Body; BODIES_COUNT]) {
    let (sun, planets) = bodies.split_first_mut().unwrap();
    sun.velocity = Default::default();
    for planet in planets {
        sun.velocity -= &planet.velocity * (planet.mass / SOLAR_MASS);
    }
}

fn compute_energy(bodies: &[Body; BODIES_COUNT]) -> f64 {
    let mut energy = 0.0;
    for (i, body1) in bodies.iter().enumerate() {
        energy += 0.5 * body1.mass * body1.velocity.sum_squares();
        for body2 in &bodies[i + 1..] {
            let d_pos = &body1.position - &body2.position;
            energy -= body1.mass * body2.mass / d_pos.sum_squares().sqrt();
        }
    }
    energy
}

fn main() {
    let turns = std::env::args()
        .nth(1)
        .and_then(|value| value.parse::<usize>().ok())
        .unwrap_or(1_000_000);
    let mut bodies = STARTING_STATE;
    offset_momentum(&mut bodies);
    let initial_energy = compute_energy(&bodies);
    let started = Instant::now();
    advance(&mut bodies, 0.01, turns);
    let seconds = started.elapsed().as_secs_f64();
    let final_energy = compute_energy(&bodies);
    println!(
        "rust-game-3,{turns},{seconds:.9},{:.3},{:.3},{initial_energy:.12},{final_energy:.12},0,0",
        seconds * 1.0e9 / turns as f64,
        turns as f64 / seconds,
    );
}
